aptos-move/move-examples/defi/sources/locked_coins.move

```diff
@@ -0,0 +1,145 @@
+module defi::locked_coins {
+    use aptos_framework::account;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::event::{Self, EventHandle};
+    use aptos_framework::timestamp;
+    use aptos_std::table::{Self, Table};
+    use std::error;
+    use std::signer;
+
+    /// Represents a lock of coins until some specified unlock time. Afterward, the recipient can claim the coins.
+    struct Lock<phantom CoinType> has store {
+        coins: Coin<CoinType>,
+        unlock_time_secs: u64,
+    }
+
+    /// Holder for a map from recipients => locks.
+    /// There can be at most one lock per recipient.
+    struct Locks<phantom CoinType> has key {
+        locks: Table<address, Lock<CoinType>>,
+        claim_events: EventHandle<ClaimEvent>,
+    }
+
+    /// Event emitted when a recipient claims unlocked coins.
+    struct ClaimEvent has drop, store {
+        recipient: address,
+        amount: u64,
+        claimed_time_secs: u64,
+    }
+
+    /// No locked coins found to claim.
+    const ELOCK_NOT_FOUND: u64 = 1;
+    /// Lockup has not expired yet.
+    const ELOCKUP_HAS_NOT_EXPIRED: u64 = 2;
+    /// Can only create one active lock per recipient at once.
+    const ELOCK_ALREADY_EXISTS: u64 = 3;
+
+    /// `Sponsor` can add locked coins for `recipient` with given unlock timestamp (in seconds).
+    /// There's no restriction on unlock timestamp so sponsors could technically add coins for an unlocked time in the
+    /// past, which means the coins are immediately unlocked.
+    public entry fun add_locked_coins<CoinType>(
+        sponsor: &signer, recipient: address, amount: u64, unlock_time_secs: u64) acquires Locks {
+        let sponsor_address = signer::address_of(sponsor);
+        if (!exists<Locks<CoinType>>(sponsor_address)) {
+            move_to(sponsor, Locks {
+                locks: table::new<address, Lock<CoinType>>(),
+                claim_events: account::new_event_handle<ClaimEvent>(sponsor),
+            })
+        };
+
+        let locks = borrow_global_mut<Locks<CoinType>>(sponsor_address);
+        let coins = coin::withdraw<CoinType>(sponsor, amount);
+        assert!(!table::contains(&locks.locks, recipient), error::already_exists(ELOCK_ALREADY_EXISTS));
+        table::add(&mut locks.locks, recipient, Lock<CoinType> { coins, unlock_time_secs });
```

Comment:
if no one is claiming the coin, this coin will be locked forever. should we provide option that claimer has to claim the coin within a time framework. beyond a time, the sponsor can re-claim the offered coin

Reply:
Good point regarding expiration. It's an interesting feature though 🤔 For now I'd keep it simple and not have that feature, but we can certainly revisit and include it in the future if the community wants to bring this module into the Aptos framework.